Add choropleth support

akacarlyann opened the pull request:
Basic implementation of choropleth map-type with ChoroplethViz and choropleth.html; addresses #4

ryanbaumann commented on mapboxgl/templates/chloropleth.html:
```diff
@@ -0,0 +1,83 @@
+    // Fly to on click
+    map.on('click', 'polygons', function(e) {
+        map.flyTo({
+            center: e.features[0].geometry.coordinates,
```
@akacarlyann on click for choropleth, we'll want to center the map on the to the mouse pointer click location since the polygon could span a very large area.

akacarlyann replied:
Addressed with d5a5435.

ryanbaumann commented on mapboxgl/templates/chloropleth.html:
```diff
@@ -0,0 +1,83 @@
+    
+    // Add data layer
+    map.addLayer({
+        "id": "polygons",
```
Let's make the layer name more descriptive- `choropleth-fill`.  Then a label layer could be `choropleth-label`

akacarlyann replied:
Addressed with d5a5435.

akacarlyann replied:
Added "chloropleth-label" layer in 3a5b3e6.

ryanbaumann commented on mapboxgl/templates/chloropleth.html:
```diff
@@ -0,0 +1,83 @@
+    style: '{{ styleUrl }}', 
+    center: {{ center }},
+    zoom: {{ zoom }},
+    transformRequest: (url, resourceType)=> {
```
We'll want to change this to the latest master branch version that only adds the URL parameter if the request is to a Mapbox API: https://github.com/mapbox/mapboxgl-jupyter/blob/master/mapboxgl/templates/circle.html#L11

akacarlyann replied:
Addressed with 0cd6da8.

ryanbaumann commented on mapboxgl/templates/chloropleth.html:
```diff
@@ -0,0 +1,83 @@
+map.on('style.load', function() {
+    
+    // Add data source
+    map.addSource("data", {
```
What if the choropleth source is a vector tile?  This would be very common when working with Pandas, for example, when you have a tabular set of data that has a geo identifier like `postcode` that you want to match to a vector tile polygon source with a feature property matching `postcode`.  https://www.mapbox.com/mapbox-gl-js/example/data-join/

ryanbaumann replied:
I'd love to figure out a way to make a `choropleth` viz in `mapboxgl` use the same python class constructor whether a user has a local geojson file or has a vector tile source to match to.

ryanbaumann commented on mapboxgl/viz.py:
```diff
@@ -268,3 +268,39 @@ def add_unique_template_variables(self, options):
+                 data,
+                 color_property=None,
+                 color_stops=None,
+                 color_default='grey',
```
We should add support for `outline-color` and `outline-width` properties here to control the style of the border around the choropleth.  

I recommend handing these options by creating a second `choropleth-line` layer, using the line type in Mapbox GL style spec, where the `line-color` and `line-width` can be controlled easily.

akacarlyann replied:
Addressed with 0cd6da8.

akacarlyann commented:
@ryanbauman Thanks for the feedback! Haven't worked with vector or raster data in the Mapbox or Leaflet environments before so I'll familiarize myself and continue working on this PR. Looks like #21 is also working on raster layers so I don't want to duplicate implementation. Maybe classes for raster or vector layers that can be stacked with the other viz types?

Are you looking for complete replication of Mapbox's example chloropleth style, hover, etc?  That would include the HTML stationary popup that calls out the color-property on hover.

ryanbaumann commented:
@akacarlyann 

> Looks like #21 is also working on raster layers so I don't want to duplicate implementation. 
Correct.  You're spot on in this PR - lets keep the choropleth viz type in `mapboxgl-jupyter` to vector data.  

> Maybe classes for raster or vector layers that can be stacked with the other viz types?

Yes, working on that feature in another issue here https://github.com/mapbox/mapboxgl-jupyter/issues/34#issuecomment-363304930

> Are you looking for complete replication of Mapbox's example chloropleth style, hover, etc? That would include the HTML stationary popup that calls out the color-property on hover.

Yes, that's a great reference starting point.  Ideally, the default choropleth would have this basic format: https://bl.ocks.org/ryanbaumann/2636bf33c7f1f6c53eca8fd34675f0bb

akacarlyann commented:
@ryanbaumann  Made some progress today wrt adding chloropleth-line and chloropleth-label layers to ChloroplethViz and its template.  Still need to add tests/docs and support for vector tile sources.

ryanbaumann commented:
Awesome progress @akacarlyann!  I'm happy to help with the vector tile source code if you can hit docs and tests for the existing functionality.

ryanbaumann commented:
@akacarlyann if you've already started on Sphynx docs, let me know.  We're planning to use a simpler approach to docs as a markdown page in the first release.  See the PR at https://github.com/mapbox/mapboxgl-jupyter/pull/46

akacarlyann commented:
@ryanbaumann I'm happy to add to the markdown development (familiar with that format) and documentation of the code I've contributed.  Haven't written up anything in Sphinx yet but I have some notes. Let me know how you me want to integrate the chloropleth  docs (fork and PR the current markdown PR?).

ryanbaumann commented:
Markdown docs are merged into master @akacarlyann 

https://github.com/mapbox/mapboxgl-jupyter/tree/master/docs-markdown

akacarlyann commented:
Great. I'll see what I can do with that this weekend :)

akacarlyann commented:
@ryanbaumann I've pushed commits for adding ChloroplethViz to the markdown docs.  I'm at the 85% solution for supporting vector tile sources.

ryanbaumann commented:
Amazing @akacarlyann.  Vector tile sources with data joins are going to be super powerful for users.  Thanks for pushing on this - let me know when you're ready for a review.

akacarlyann commented:
@ryanbaumann There is no polygon data in the tests/ module.  Do you have a polygon dataset that fits well with the existing tests?  Otherwise I'll build in something like 
```
@pytest.fixture()
def polygon_data():
    with open('tests/polygons.geojson') as fh:
        return json.loads(fh.read())
```
probably using the US states population density info I've been working with.

ryanbaumann commented:
@akacarlyann that's excellent re: using a small local tileset like US States.  Natural Earth is a great source of simplified small polygon boundaries for testing - http://www.naturalearthdata.com/downloads/110m-physical-vectors/

ajfriend commented:
This is awesome work! I'll use it immediately.

And, apologies for the random nit from the internet, but as far as naming goes, what's the difference between "chloropleth" and "choropleth" (which I've seen more often)?

akacarlyann commented:
@ajfriend Ha! The difference is mine have been the only eyes on this so far and now I'm just going to go update spelling throughout. Thanks for the comment!

akacarlyann commented:
@ajfriend Ha! I think the main difference is mine have been the only eyes on this PR so far.  Seems like Mapbox uses the Choropleth spelling (how did I miss that?), so I'll probably update spelling, though I originally found Chloropleth at https://python-graph-gallery.com/chloropleth-map/  .  Thanks for your comment!

Juanlu001 commented:
The choropleth notebook included in this PR is giving me some sort of formatting errors. I cannot open it with nbviewer either:

http://nbviewer.jupyter.org/github/akacarlyann/mapboxgl-jupyter/blob/ac8c1000833e56cd7d806c56894e5df2f694aa45/examples/choropleth-viz-example.ipynb

Juanlu001 commented:
Trivial fix, probably due to find 'n' replace

```diff
--- a/examples/choropleth-viz-example.ipynb
+++ b/examples/choropleth-viz-example.ipynb
@@ -44,9 +44,7 @@
     "    [10000.0,'rgb(128,0,38)']\n",
     "]\n",
     "   \n",
-    "viz = 
-
-ChoroplethViz('us-states.geojson', \n",
+    "viz = ChoroplethViz('us-states.geojson', \n",
     "                     color_property='density',\n",
     "                     color_stops=sample_color_stops,\n",
     "                     color_function_type='interpolate',\n",
@@ -80,9 +78,7 @@ ChoroplethViz('us-states.geojson', \n",
     "    ['California', 'rgb(142,68,173)'],\n",
     "]\n",
     "    \n",
-    "viz = 
-
-ChoroplethViz('us-states.geojson', \n",
+    "viz = ChoroplethViz('us-states.geojson', \n",
     "                     color_property='name', \n",
     "                     color_stops = match_color_stops, \n",
     "                     color_function_type = 'match', \n",
```

ryanbaumann commented:
cc @akacarlyann 👆

akacarlyann commented:
@ryanbaumann I've found and been able to implement examples of the data-join technique that rely on generating colors based on a numerical property (index, density, unemployment rate...). This method is problematic for __ChoroplethViz__ because colors generated in this way will override the user-defined __color_stops__ argument and cause a mismatch with the map's legend. 
- https://www.mapbox.com/mapbox-gl-js/example/data-join/
- [https://bl.ocks.org/ryanbaumann/.../](https://bl.ocks.org/ryanbaumann/63ca44169e657d66c1878f1083017e85/99a0a99cb7ca748dc4a74a79e82761c34dcce1c8)

Looks like you were involved in all of these!

I'm most interested in this map:  
https://www.mapbox.com/labs/dataset-join/#3.03/36.45/-90.02  
https://blog.mapbox.com/visualize-the-usas-economic-recovery-with-client-side-data-joins-2aeab23ee9b0  
Can you provide some insight into the actual color calculation function for this example?  Is this still the categorical property function method?  The color-stops in the legend lead me to believe that it _might_ be possible to preserve our current color-stop paradigm.

TL;DR I want a choropleth visualization made with either a geojson or vector polygon source to result in identically colored maps.  Thoughts on that?

ryanbaumann commented:
Great question @akacarlyann - are your latest changes pushed to this branch?  Happy to do a PR review to help recommend an approach.

akacarlyann commented:
@ryanbaumann I've pushed a couple of options (this makes me nervous since I don't want anyone to take this as a finished product).  My sense is that we would hope to generate a single color function for either vector+data-join or the current color_stops paradigm.  

With so many additional arguments only necessary for if the polygons come from a vector layer, I'll obviously need to design some new tests and/or set up more robust defaults in case one is missed.  I didn't invest any time in this yet, as I can still see this branch going in several somewhat different directions.  

Thanks in advance for taking a look!  I've updated the example notebook at examples/choropleth-viz-example.ipynb to include two samples using vector source for the polygons.

ryanbaumann commented:
Excellent @akacarlyann.  No need to worry about the PR being in an evaluation state still - I'll get you feedback by the end of the week.  Thank you!

ryanbaumann commented on mapboxgl/templates/choropleth_alt.html:
```diff
@@ -0,0 +1,204 @@
+        });
+    {% endif %}
+
+    {% if vectorPolygonSource %}
```
@akacarlyann this makes sense to me - define two different source initialization templates, and pick which one to use based on if the source is a vector tile source or geojson polygon source.

akacarlyann replied:
Sounds good.  Settled on this in b11c006.

ryanbaumann commented on mapboxgl/viz.py:
```diff
@@ -345,3 +360,88 @@ def add_unique_template_variables(self, options):
+                 vector_polygon_source=False,
+                 vector_url=None,
+                 vector_layer_name=None,
+                 vector_join_color_property=None,
```
Could we merge this with the existing `color_property` class variable?

akacarlyann replied:
I found this necessary for dealing with vector data that might have different property names than your joined geojson. For example, you have the US States vector source, which has a property called STATEFP, which is the "id" and our geojson object, with a property called "id". If there is a way to override vector property names I'd love to hear it!! Rather than rely on users to always have well-formatted geojson, I thought this would provide more flexibility.

ryanbaumann commented on mapboxgl/viz.py:
```diff
@@ -345,3 +360,88 @@ def add_unique_template_variables(self, options):
+        """Update map template variables specific to heatmap visual"""
+
+        # set line stroke dash interval based on line_stroke property
+        if self.line_stroke == "dashed":
```
Nice.  Very nice for data viz use cases to just specify the type of viz they want instead of writing style json code or Mapbox expressions 👍

ryanbaumann commented on mapboxgl/viz.py:
```diff
@@ -345,3 +360,88 @@ def add_unique_template_variables(self, options):
+                 vector_url=None,
+                 vector_layer_name=None,
+                 vector_join_color_property=None,
+                 join_data=[],
```
Could this be merged with the class variable `data`?

akacarlyann replied:
I did start this way initially but it added significant complexity/logic to the Jinja template. Seems like a good option if have two separate choropleth templates though!

ryanbaumann replied:
Yeah.  If it adds significant complexity to merge these into one, let's separate the `geojson choropleth` from the `vectortile choropleth` into two classes.  I'll leave this up to your judgment @akacarlyann

akacarlyann replied:
With the separate html templates, this turned out to be simple so I've merged join_data with data in b11c006.

ryanbaumann commented on mapboxgl/viz.py:
```diff
@@ -345,3 +360,88 @@ def add_unique_template_variables(self, options):
+
+    def __init__(self,
+                 data,
+                 vector_polygon_source=False,
```
Could we automatically sniff this based on if a specific class variable is not `None`, such as `vector_url`?

akacarlyann replied:
Sure thing.

akacarlyann replied:
Took care of this in b11c006.

ryanbaumann commented on mapboxgl/templates/choropleth_alt.html:
```diff
@@ -0,0 +1,204 @@
+
+        // create categorical color stops for features in vector layer
+        joinData.forEach(function(row, index) {
+            var red = Math.max(255 - row["{{ colorProperty }}"], 0),
```
@akacarlyann We should try to generate the data-join match expression in the Python viz class here, instead of in Javascript. This should allow us to re-use the color palette helper functions in the mapboxgl.Colors class, and limit the computation on the javascript side to just loading the join data to create popup values for each property you join from your data set to the polygon feature.

akacarlyann replied:
Agreed, but I still haven't worked out a way that would exactly replicate the color assignment you would get if the map were just styling points with the same values using an interpolate function.

ryanbaumann replied:
Ah.  I recommend chroma.js for color interpolation -> https://github.com/gka/chroma.js/.  This should be the equivalent in Python for color interpolation -> https://pypi.python.org/pypi/Chroma/0.2.0

I'd prefer if we handle this on the Python side to separate business logic concers to that code base, as opposed to the Jinja template JS files.  The rough steps are:

1. Create a chroma color scale from a color list and domain of stops
2. Calculate the value of color to use for each unique stop
3. Create a mapboxgl `match` expression with the `id` of each feature and `color` calculated from the scale of each color.

Does that make sense?  I think we could wrap up all of 1->3 in one or two helper functions in `mapboxgl.utils`.

akacarlyann replied:
Can these helper functions be part of a different PR?

ryanbaumann commented on mapboxgl/templates/choropleth_alt.html:
```diff
@@ -0,0 +1,204 @@
+
+calcCircleColorLegend({{ colorStops }}, "{{ colorProperty }}");
+
+function calcNumericColorValue(num, arr) {
```
Can we remove this helper function if we generate the color interpolation in the Python thread (see comment below)?

akacarlyann replied:
That's cruft. It's not being used, sorry!

ryanbaumann commented on mapboxgl/templates/choropleth_alt.html:
```diff
@@ -0,0 +1,204 @@
+    }
+});
+
+var joinData = {{ joinData }};
```
Since the number of stops we generate will probably be less than 10k or so, this won't be TOO slow.  Ideally, we would allow the user to load this data async from a local file HTTP request, just like they can for a normal `data` object as a geojson source.

Not critical to hit for this PR

akacarlyann replied:
Yep I'd really like to get this in as well for consistency.  Is there a way to iterate through geojson added to the map as a source?

ryanbaumann replied:
@akacarlyann yes, for geojson sources only, you can call `map.getSource('data')._data` to get a copy of the source geojson data.

ryanbaumann replied:
However, the `join data` in this case will not be a geojson source, no?  The polygons will be in a vector tile source, and all the join data will contain is a JSON format set of `key: value` pairs, used to create the `match` expression for the style.

akacarlyann replied:
In this case, are you suggesting that the __data__ param can be either geojson or json?  Won't that get confusing?

ryanbaumann commented on examples/choropleth-viz-example.ipynb:
```diff
@@ -90,6 +92,152 @@
+    "    [5000.0,'rgb(128,0,38)']\n",
+    "]\n",
+    "\n",
+    "# some pre-computed colors (avoid computing colors in JavaScript template)\n",
```
Once we have the class finalized, let's generate this data-join color stops with a helper function instad of manually.

ryanbaumann commented:
@akacarlyann added replies to your comments above.  It sounds like the biggest decision points right now are to:

1. Whether or not to separate out a mapboxgl.viz class for a `choropleth` viz from a vector tile source vs. a geojson source.
    * My preference here is no, as it would cause class sprawl.  We'd follow this same paradigm for other viz types too, like extrusions and linestrings.
    * If we use the same class for both geojson and vector sources, whether or not to create new style properties to the class specifically for the vector source vs. the geojson source.
2. How to calculate the data-join match between vector tiles and local data in Python - should it be in Python or in JS?
    * My preference here is to do this in Python using the Chroma library.  

Let me know your thoughts on the approach to ^^

ryanbaumann commented:
@akacarlyann are you planning on wrapping up this PR, or would you like some assistance?

ryanbaumann commented:
This is looking amazing, @akacarlyann!  Great job coalescing the various different geojson vs. vector tile methods of creating fill maps into one `ChoroplethViz` syntax.  I took a quick look and pushed a small doc update to keep track of the syntax - all looks good.

Final items to hit before merge that I'm sure you're tracking:

- [ ] "Use  [chroma.js scales](https://gka.github.io/chroma.js/#color-scales) to calculate color values for the `match` style from the user-specified color stops
- [ ] Update tooltips to reference data passed to the visual, in addition to properties in the vector tiles.
- [ ] Update the vector tile join example to use ^^

I think that's it - nice work!

akacarlyann commented:
Thanks! The tooltips do append the joined data property. Are you wanting additional features to also migrate to the vector features tooltips? If so that's not too much to add, just didn't realize it was in the scope.

And yep still on the color matching. I'll push an update later today most likely.

ryanbaumann commented:
Excellent @akacarlyann !  

RE: 
> The tooltips do append the joined data property. Are you wanting additional features to also migrate to the vector features tooltips? If so that's not too much to add, just didn't realize it was in the scope.

They should, I but they wern't showing up running the example notebook.  I narrowed it down to two items to fix - added comments to your latest commit.

ryanbaumann commented on mapboxgl/templates/vector_choropleth.html:
```diff
@@ -0,0 +1,104 @@
+    
+    {% block choropleth %}
+
+    var joinData = {{ joinData }};
```
We don't want to store the raw data, since we re-map this into an object for the tooltip and array for the style below.  Scope this as a function and import the variable in the function scope using `let joinData = {{ joinData }}`.

ryanbaumann commented on mapboxgl/templates/vector_choropleth.html:
```diff
@@ -0,0 +1,104 @@
+
+    {% if joinData %}
+        var vectorColorStops = [],
+            popUpKeys = [];
```
Store popup keys in an object `{}` for quick retrieval by key.

ryanbaumann commented on mapboxgl/templates/vector_choropleth.html:
```diff
@@ -0,0 +1,104 @@
+
+    {% endblock choropleth %}
+
+    {% block choropleth_popup %}
```
When I run this example code, the mousemove function below is not applied to the map I believe because the `choropleth_popup` template in `choropleth.html` is being executed instead of this block.  I'm sure this could be addressed multiple ways, but I confirmed a quick fix by adding the same block name `choropleth_popup` to both `choropleth.html` and `vector_choropleth.html` with the complete code for the popup specific to that template in each.

akacarlyann replied:
I think I've covered this in 722d747, but I wasn't able to replicate the original problem.  Can you confirm if it's working now?

akacarlyann commented:
@ryanbaumann  I finally got the color logic working over the weekend!  I'm cleaning up my last commits now.

ryanbaumann commented:
Amazing @akacarlyann, thank you!  Let me know when you're ready for a review.

akacarlyann commented:
@ryanbaumann It's ready for review now :) I took a stab at tests for the new functions I introduced but I need a little help getting those on track. I also noticed when merging master back in to my branch that there is an import from __colour__ that conflicts with mine from __chroma-py__, so I renamed one of the __Color__ classes. I'm happy to refactor if these serve similar purposes, just let me know!

ryanbaumann commented:
Awesome - I'll get you a PR review by Saturday, thanks @akacarlyann

ryanbaumann commented on mapboxgl/templates/vector_choropleth.html:
```diff
@@ -35,7 +15,7 @@
-            "fill-color": generatePropertyExpression("match", "{{ vectorJoinColorProperty }}", vectorColorStops, "{{ defaultColor }}"), 
+            "fill-color": generatePropertyExpression("match", "{{ vectorJoinColorProperty }}", {{ vectorColorStops }}, "{{ defaultColor }}"), 
             "fill-opacity": {{ opacity }}
         }
```
Add layer filter here (see comment on how to create)

ryanbaumann commented on mapboxgl/templates/vector_choropleth.html:
```diff
@@ -79,9 +59,23 @@
+
+{% block choropleth_popup %}
+
+    // extract JSON property used for data-driven styling to add to popup
```
Move this section to create the popup join data before the layer creation in the `map.on(style.load) function.  We'll want to generate a `filter` expression here too that limits the vector features visible to ones that match data in the user's data.

Here's how I'd recommend doing it:

```javascript
{% block choropleth %}
    
    // extract JSON property used for data-driven styling to add to popup
    {% if joinData %}

        let joinData = {{ joinData }};
        # Create filter for layers from join data
        let layerFilter = ['in', "{{ vectorJoinColorProperty }}"]
        var popUpKeys = {};

        joinData.forEach(function(row, index) {
            popUpKeys[row["{{ dataJoinProperty }}"]] = row["{{ colorProperty }}"];
            layerFilter.push(row["{{ dataJoinProperty }}"])
        });

    {% endif %}

    // Add vector data source
    map.addSource("vector-data", {
        type: "vector",
        url: "{{ vectorUrl }}",
    });

    // Add layer from the vector tile source with data-driven style
    map.addLayer({
        "id": "choropleth-fill",
        "type": "fill",
        "source": "vector-data",
        "source-layer": "{{ vectorLayer }}",
        "paint": {
            "fill-color": {
                "type": "categorical",
                "property": "{{ vectorJoinColorProperty }}", 
                "stops": {{ vectorColorStops }}, 
                "default": "{{ defaultColor }}"
            },
            "fill-opacity": {{ opacity }}
        },
        filter: layerFilter
    }, "{{ belowLayer }}");

    // Add border layer
    map.addLayer({
        "id": "choropleth-line",
        "source": "vector-data",
        "source-layer": "{{ vectorLayer }}",
        "type": "line",
        "layout": {
            "line-join": "round",
            "line-cap": "round"
        },
        "paint": {
            {% if lineDashArray %}
                "line-dasharray": {{ lineDashArray }},
            {% endif %}
            "line-color": "{{ lineColor }}",
            "line-width": {{ lineWidth }},
            "line-opacity": {{ opacity }}
        },
        filter: layerFilter
    }, "{{ belowLayer }}" );

    // Add label layer
    map.addLayer({
        "id": "choropleth-label",
        "source": "vector-data",
        "source-layer": "{{ vectorLayer }}",
        "type": "symbol",
        "layout": {
            {% if labelProperty %}
            "text-field": "{{ labelProperty }}",
            {% endif %}
            "text-size" : generateInterpolateExpression('zoom', [[0,8],[22,16]] ),
            "text-offset": [0,-1]
        },
        "paint": {
            "text-halo-color": "white",
            "text-halo-width": 1
        },
        filter: layerFilter
    }, "{{belowLayer}}" );

{% endblock choropleth %}
```

ryanbaumann commented on mapboxgl/templates/vector_choropleth.html:
```diff
@@ -79,9 +59,23 @@
+    {% endif %}
 
-    {% block choropleth_popup %}
     // Show the popup on mouseover
```
Add in the logic to remove the popup here on mouseleave.

```javascript
    map.on('mouseleave', 'choropleth-fill', function() {
        map.getCanvas().style.cursor = '';
        popup.remove();
    });
    
    // Fly to on click
    map.on('click', 'choropleth-fill', function(e) {
        map.flyTo({
            center: e.lngLat,
            zoom: map.getZoom() + 1
        });
    });
```

akacarlyann commented:
@ryanbaumann Thanks, I'll get those in today :)

akacarlyann commented:
Alrighty, @ryanbaumann , tests and your requested changes ready for review!

ryanbaumann commented:
@akacarlyann this is so good.  Thanks for pushing through a difficult PR, laying the foundation for all sorts of new viz types with the data-join technique packaged up!
